File: runners/csharp/Kikker.Billing.IntegrationTests/Helpers/WaitHelper.cs
namespace Kikker.Billing.IntegrationTests.Helpers;

internal static class WaitHelper
{
    public static async Task WaitUntilAsync(Func<Task<bool>> condition, TimeSpan timeout, TimeSpan pollInterval, string timeoutMessage)
    {
        var startedAt = DateTime.UtcNow;
        while (DateTime.UtcNow - startedAt < timeout)
        {
            if (await condition())
            {
                return;
            }

            await Task.Delay(pollInterval);
        }

        throw new TimeoutException(timeoutMessage);
    }
}

File: runners/java/src/test/java/com/kikker/billing/integrationtests/helpers/WaitHelper.java
package com.kikker.billing.integrationtests.helpers;

import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.Callable;

public final class WaitHelper {
    private WaitHelper() {
    }

    public static void waitUntil(Callable<Boolean> condition, Duration timeout, Duration pollInterval, String timeoutMessage)
        throws Exception {
        Instant startedAt = Instant.now();
        while (Duration.between(startedAt, Instant.now()).compareTo(timeout) < 0) {
            if (condition.call()) {
                return;
            }

            Thread.sleep(pollInterval.toMillis());
        }

        throw new IllegalStateException(timeoutMessage);
    }
}
